src/java.desktop/share/classes/sun/font/StrikeCache.java

```diff
@@ -98,40 +105,161 @@ public final class StrikeCache {
-
-    /* Native method used to return information used for unsafe
-     * access to native data.
-     * return values as follows:-
-     * arr[0] = size of an address/pointer.
-     * arr[1] = size of a GlyphInfo
-     * arr[2] = offset of advanceX
-     * arr[3] = offset of advanceY
-     * arr[4] = offset of width
-     * arr[5] = offset of height
-     * arr[6] = offset of rowBytes
-     * arr[7] = offset of topLeftX
-     * arr[8] = offset of topLeftY
-     * arr[9] = offset of pixel data.
-     * arr[10] = address of a GlyphImageRef representing the invisible glyph
-     */
-    static native void getGlyphCacheDescription(long[] infoArray);
+    static final int nativeAddressSize = (int)ValueLayout.ADDRESS.byteSize();
+    static final long invisibleGlyphPtr = getInvisibleGlyphPtr(); // a singleton.
+
+    static native long getInvisibleGlyphPtr();
+
+    public static final StructLayout GlyphImageLayout = MemoryLayout.structLayout(
+        JAVA_FLOAT.withName("xAdvance"), // 0+4=4,
+        JAVA_FLOAT.withName("yAdvance"), // 4+4=8,
+        JAVA_CHAR.withName("width"),     // 8+2=10,
+        JAVA_CHAR.withName("height"),    // 10+2=12
+        JAVA_CHAR.withName("rowBytes"),  // 12+2=14
+        JAVA_BYTE.withName("managed"),   // 14+1=15
+        MemoryLayout.paddingLayout(1),   // 15+1=16
+        JAVA_FLOAT.withName("topLeftX"), // 16+4=20
+        JAVA_FLOAT.withName("topLeftY"), // 20+4=24
+        ADDRESS.withName("cellInfo"),    // 24+8=32
+        ADDRESS.withName("image")        // 32+8=40
+     );
+
+   private static final long GLYPHIMAGESIZE = GlyphImageLayout.byteSize();
+
+   private static VarHandle getVarHandle(StructLayout struct, String name) {
+        VarHandle h = struct.varHandle(PathElement.groupElement(name));
+        /* insert 0 offset so don't need to pass arg every time */
+        return MethodHandles.insertCoordinates(h, 1, 0L).withInvokeExactBehavior();
+    }
+
+    private static final VarHandle xAdvanceHandle = getVarHandle(GlyphImageLayout, "xAdvance");
+    private static final VarHandle yAdvanceHandle = getVarHandle(GlyphImageLayout, "yAdvance");
+    private static final VarHandle widthHandle    = getVarHandle(GlyphImageLayout, "width");
+    private static final VarHandle heightHandle   = getVarHandle(GlyphImageLayout, "height");
+    private static final VarHandle rowBytesHandle = getVarHandle(GlyphImageLayout, "rowBytes");
+    private static final VarHandle managedHandle  = getVarHandle(GlyphImageLayout, "managed");
+    private static final VarHandle topLeftXHandle = getVarHandle(GlyphImageLayout, "topLeftX");
+    private static final VarHandle topLeftYHandle = getVarHandle(GlyphImageLayout, "topLeftY");
+    private static final VarHandle cellInfoHandle  = getVarHandle(GlyphImageLayout, "cellInfo");
+    private static final VarHandle imageHandle    = getVarHandle(GlyphImageLayout, "image");
+
+    @SuppressWarnings("restricted")
+    static final float getGlyphXAdvance(long ptr) {
```

Comment:
now, I'm not an expert of this code, but I notice that you have accessors that seem to take a bare `long` instead of `MemorySegment`. Have you tried pushing segments deeper in the implementation? That way I think you could completely auto-generate this code using jextract. (of course what you have is not bad - I'm mostly trying to see if there's a way to get there w/o all these calls to `MemorySegment::reinterpret`).

Reply:
To be clear - I'm assuming the `ptr` parameter comes from a single native call. That call returns a pointer to a struct, or maybe an array of structs. If so, we could reinterpret right after the native call, with the right size (the number of structs returned by the native call), and then all the accessors will be just "plain" memory segment accessors, no need to reinterpret (as the segment will already have the correct size).

Reply:
Each pointer refers to a single struct, not an array.
In this fix I am focused on not using Unsafe.
I don't want the change to go further than that.
The 'long's are retrieved/ passed around/ back down to native in lots of places so it would turn a  focused change into a much bigger change.
grep in java.desktop for "getGlyphImage" to see the tip of the iceberg.